### PR TITLE
[CONTINT-4105] Support arbitrary container-ids to collect container metrics

### PR DESCRIPTION
**Pull request by AliDatadog**
### What does this PR do?
This PR adds support of arbitrary container-id for containerd. We now can collect their container metrics and tags.

### Motivation
Reduce the number of false negatives with a more robust solution to retrieve container-ids.

### Additional Notes
[RFC](https://datadoghq.atlassian.net/wiki/spaces/CONT/pages/3695837463/Support+arbitrary+container-ids+to+collect+container+metrics) 

### Describe how to test/QA your changes
Deploy the agent on a kind cluster.
Pull an image on the node with `docker exec <node-id> ctr i pull docker.io/library/redis:latest`
Start a redis container on the node with `docker exec <node-id> ctr run docker.io/library/redis:latest redis`
Make sure container metrics can be found with the right container-id (redis).
[Notebook](https://dddev.datadoghq.com/notebook/8186362/ali-may-13-2024-12-25?cell-am49macm-live=false&cell-am49macm-range=87758&cell-am49macm-start=1715593106106).

### CODE REVIEW

**Comment by AliDatadog** [comment]
/trigger-ci --variable RUN_ALL_BUILDS=true --variable RUN_KITCHEN_TESTS=true --variable RUN_E2E_TESTS=on --variable RUN_UNIT_TESTS=on --variable RUN_KMT_TESTS=on

File: releasenotes/notes/support-arbitrary-container-id-cc0efdf7c156b7ad.yaml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+fixes:
+  - |
+    Fix a bug where container metrics and container tags were not being
+    collected for containers with arbitrary container IDs with containerd.
```
Comment (buraizu) [inline releasenotes/notes/support-arbitrary-container-id-cc0efdf7c156b7ad.yaml:4]:
```suggestion
    Fix a bug where containerd container metrics and container tags were not being
    collected for containers with arbitrary container IDs.
```

**Review by buraizu — APPROVED** [review]
Approving with a minor suggestion to the release note

File: pkg/util/containers/metrics/system/filter_container.go [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+		wlm:  wlm,
+	}
+	return cf
+}
+
+func (cf *containerFilter) start() {
+	if cf.wlm == nil {
+		return
+	}
+	evBundle := cf.wlm.Subscribe("cid-mapper", workloadmeta.NormalPriority, workloadmeta.NewFilter(
```
Comment (vboulineau) [inline pkg/util/containers/metrics/system/filter_container.go:127]: Do we need to actually subscribe? If you create a `ContainerFilter` with access to WLM, you could:
- Check if it matches the regexp
- If it does not match the regexp, you could check if it's a leaf (we'd need to check what we can do on the parsing on that front) and call `WLM.GetContainer`.

File: pkg/util/containers/metrics/system/filter_container.go [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+		&workloadmeta.FilterParams{
+			Kinds:     []workloadmeta.Kind{workloadmeta.KindContainer},
+			Source:    workloadmeta.SourceAll,
+			EventType: workloadmeta.EventTypeAll,
+		},
+	))
+	defer cf.wlm.Unsubscribe(evBundle)
```
Comment (vboulineau) [inline pkg/util/containers/metrics/system/filter_container.go:134]: This is actually never called

File: pkg/util/containers/metrics/system/filter_container.go [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+	}
+}
+
+// ContainerFilter returns a filter that will match cgroup folders containing a container id
+func (cf *containerFilter) ContainerFilter(path, name string) (string, error) {
+	if res, _ := cgroups.ContainerFilter(path, name); res != "" {
+		return res, nil
+	}
+	cf.mutex.RLock()
+	res := cf.trie.Get(path)
```
Comment (vboulineau) [inline pkg/util/containers/metrics/system/filter_container.go:169]: As we have path matching, is the trie actually useful compared to a map?

File: pkg/util/containers/metrics/system/filter_container.go [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+	root *trieNode
+}
+
+func newSuffixTrie() *suffixTrie {
+	return &suffixTrie{
+		root: newTrieNode(),
+	}
+}
+
+func reverse(s string) string {
```
Comment (vboulineau) [inline pkg/util/containers/metrics/system/filter_container.go:39]: By reversing, you target matching optimization by expecting path end is more unique than path start (which is true), however it means you basically don't benefit from memory savings of the trie as you're going to store all paths starts as many times as entries (mostly).

**Review by vboulineau — COMMENTED** [review]
Note that the implementation is racy by nature, as we depend on the subscriber to have done the work when `ContainerFilter` is called.
It will normally always converge after few seconds, but it should be noted.

File: pkg/util/containers/metrics/system/collector_linux.go [hunk 6]
```diff
@@ -61,11 +61,18 @@ func newSystemCollector(cache *provider.Cache) (provider.CollectorMetadata, erro
 		hostPrefix = "/host"
 	}
 
+	var w workloadmeta.Component
+	unwrapped, ok := wlm.Get()
+	if ok {
+		w = unwrapped
```
[the remark this answers is not in the document]
Comment (AliDatadog) [reply pkg/util/containers/metrics/system/collector_linux.go:67]: In this case it's intentional. The `start` function returns nothing and the `ContainerFilter` will only call `cgroups.ContainerFilter`

File: pkg/util/containers/metrics/system/filter_container.go (hunk 3 above)
Comment (AliDatadog) [reply pkg/util/containers/metrics/system/filter_container.go:134]: I thought that the channel could be closed by workloadmeta on shutdown. Removed the Unsubscribe for now.

File: pkg/util/containers/metrics/system/filter_container.go (hunk 4 above)
Comment (AliDatadog) [reply pkg/util/containers/metrics/system/filter_container.go:168]: As discussed, `ContainerFilter` is called with a full path while the workloadmeta object stores suffixes so we need to do suffix matching. I'll improve the doc and split the files.

File: pkg/util/containers/metrics/system/filter_container.go [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+
+import (
+	"sync"
+
+	"github.com/DataDog/datadog-agent/comp/core/workloadmeta"
+	"github.com/DataDog/datadog-agent/pkg/util/cgroups"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+type trieNode struct {
```
[the remark this answers is not in the document]
Comment (AliDatadog) [reply pkg/util/containers/metrics/system/filter_container.go]: Moved it to its own package and made it more generic

**Review by hush-hush — APPROVED** [review]

**Review by vickenty — APPROVED** [review]

**Review by mellon85 — APPROVED** [review]

**Review by robertjli — APPROVED** [review]

**Review by wdhif — APPROVED** [review]

**Review by ajgajg1134 — APPROVED** [review]

**Review by vboulineau — APPROVED** [review]

**Pull request merged**